test/gtest/type-builder.cpp

```diff
@@ -1095,6 +1121,256 @@ FUZZ_TEST(TypeFuzzTest, TestHeapTypeRelationsFuzz)
 
 #endif // FUZZTEST
 
+TEST_F(TypeTest, TestTypeRelations) {
+  Type any = Type(HeapType::any, NonNullable, Inexact);
+  Type nullAny = Type(HeapType::any, Nullable, Inexact);
+  Type exactAny = Type(HeapType::any, NonNullable, Exact);
+  Type nullExactAny = Type(HeapType::any, Nullable, Exact);
+
+  HeapType defined = Struct();
+  Type def = Type(defined, NonNullable, Inexact);
+  Type nullDef = Type(defined, Nullable, Inexact);
+  Type exactDef = Type(defined, NonNullable, Exact);
+  Type nullExactDef = Type(defined, Nullable, Exact);
+
+  Type none = Type(HeapType::none, NonNullable, Inexact);
+  Type nullNone = Type(HeapType::none, Nullable, Inexact);
+  Type exactNone = Type(HeapType::none, NonNullable, Exact);
+  Type nullExactNone = Type(HeapType::none, Nullable, Exact);
+
+  Type func = Type(HeapType::func, NonNullable, Inexact);
+  Type nullFunc = Type(HeapType::func, Nullable, Inexact);
+  Type exactFunc = Type(HeapType::func, NonNullable, Exact);
+  Type nullExactFunc = Type(HeapType::func, Nullable, Exact);
```

Comment:
What does `(exact (ref func))` mean? No actual entity can have exactly the type of a basic type, so is it uninhabitable?

Reply:
Yes, `(ref exact func)` is uninhabitable. It's still a valid type, though!

Reply:
That should be safe to leave for later. It looks like it's only used by the fuzzer right now.

Reply:
It is used in `evaluateCastCheck()` in the optimizer, but maybe it would just miss opportunities, not miscompile.